src/auxiliary/mpi_arrays.jl

```diff
@@ -0,0 +1,152 @@
+
+
+# Custom interface and general Base interface not covered by other parts below
+Base.parent(u::TrixiMPIArray) = u.u_local
+Base.resize!(u::TrixiMPIArray, new_size) = resize!(parent(u), new_size)
+
+mpi_comm(u::TrixiMPIArray) = u.mpi_comm
+mpi_rank(u::TrixiMPIArray) = u.mpi_rank
+# TODO: MPI. What about the following interface functions?
+# mpi_nranks(u::TrixiMPIArray) = MPI_SIZE[]
+# mpi_isparallel(u::TrixiMPIArray) = MPI_IS_PARALLEL[]
+
+
+# Implementation of the abstract array interface of Base
+# See https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array
+Base.size(u::TrixiMPIArray) = size(parent(u))
+Base.getindex(u::TrixiMPIArray, idx) = getindex(parent(u), idx)
+Base.setindex!(u::TrixiMPIArray, v, idx) = setindex!(parent(u), v, idx)
+Base.IndexStyle(::Type{TrixiMPIArray{T, N, Parent}}) where {T, N, Parent} = IndexStyle(Parent)
+Base.similar(u::TrixiMPIArray, ::Type{S}, dims::NTuple{N, Int}) where {S, N}	= TrixiMPIArray(similar(parent(u), S, dims))
+Base.axes(u::TrixiMPIArray)	= axes(parent(u))
+
+
+# Implementation of the strided array interface of Base
+# See https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-strided-arrays
+Base.strides(u::TrixiMPIArray) = strides(parent(u))
+Base.unsafe_convert(::Type{Ptr{T}}, u::TrixiMPIArray{T}) where {T} = Base.unsafe_convert(Ptr{T}, parent(u))
+Base.elsize(::Type{TrixiMPIArray{T, N, Parent}}) where {T, N, Parent} = elsize(Parent)
+
+
+# TODO: MPI. Do we need customized broadcasting? What about FastBroadcast.jl and
+#            threaded execution with `@.. thread=true`?
+# See https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting
+
+
+# Implementation of methods from ArrayInterface.jl for use with
+# LoopVectorization.jl etc.
+# See https://juliaarrays.github.io/ArrayInterface.jl/stable/
+ArrayInterface.parent_type(::Type{TrixiMPIArray{T, N, Parent}}) where {T, N, Parent} = Parent
+
+
+# TODO: MPI. Do we need LinearAlgebra methods such as `norm` or `dot`?
+
+
+# `mapreduce` functionality from Base using global reductions via MPI communication
+function Base.mapreduce(f::F, op::Op, u::TrixiMPIArray; kwargs...) where {F, Op}
+  local_value = mapreduce(f, op, parent(u); kwargs...)
+  return MPI.Allreduce(local_value, op, mpi_comm(u))
+end
+
+
+# TODO: MPI. Default settings of OrdinaryDiffEq etc.
+# Interesting options could be
+# - ODE_DEFAULT_UNSTABLE_CHECK
+# - ODE_DEFAULT_ISOUTOFDOMAIN
+# - ODE_DEFAULT_NORM
+# See https://github.com/SciML/DiffEqBase.jl/blob/master/src/common_defaults.jl
+
+# Problems and inconsistencies with TrixiMPIArrays
+# A basic question is how to handle `length`. We want `TrixiMPIArray`s to behave
+# like regular `Array`s in most code, e.g., when looping over an array (which
+# should use `eachindex`). At the same time, we want to be able to use adaptive
+# time stepping using error estimates in OrdinaryDiffEq.jl. There, the default
+# norm `ODE_DEFAULT_NORM` is the one described in the book of Hairer & Wanner,
```

Comment:
Just thinking - can we avoid this `local_length` issue if we define a `norm` function that works in parallel? That might be an alternative to having to remember to use `local_length`.

A potential downside of `local_length` - that I just noticed - is that it allows users to create code that works in serial but may fail in spectacularly surprising ways if run in parallel. That is, if someone uses `length` where `local_length` is required, it works fine in serial but may cause weird issues in parallel (especially if running with --check-bounds=no).

Reply:
Yeah, that's the issue of the minimally invasive approach using a global `length`. However, i would argue that users should better use `eachindex` in most cases, which is fine.

Reply:
No, I agree - `eachindex` should be used where possible. It makes, however, for difficult-to-understand errors, and the "wrong" use of `length` might be hard to spot in reviews. I suggest to continue making it work, but then we should revisit this (or at least capture it in an issue).

Reply:
Another alternative would be to write our own norm function and pass that as `solve(ode, alg; kwargs..., internalnorm=our_new_norm_function)`. However, that requires yet another keyword argument we need.

Reply:
Right, especially since it would fail very late during the initialization (or even worse, just hang) if forgotten. Maybe we need our own `trixi_solve` that passes some default options to OrdinaryDiffEq.jl's `solve`?

Reply:
That might be better. We don't need to solve this right now, though, do we? Maybe we just copy the current discussion to an issue and deal with it later, once we have some more experience with the new type.

Reply:
Yeah, sounds good to me. I'll leave this thread open and we can continue the discussion later.